### DISCUSSION TITLE
Add a way to have "fallback" items

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#3945] [Feature Requests]
This is especially useful in translations. If you have incomplete translations for an item, you want it to fallback the empty fields to the "next" translation in line, for example:

Dutch-België
title: Hallo wereld!
description: _empty_

Dutch-Nederland
title: Hallo daar!
description: Welkom op mijn website

should result in

title: Hallo wereld!
description: Welkom op mijn website.

This could be potentially be done with a `fallback` query parameter that will merge all nested o2m items into a single object, where nulls fallback to the next in line:

```js
{
  translations: [
    { title: 'Hallo wereld!', description: null, language: 'nl-BE' },
    { title: 'Hallo daar!', description: 'Welkom op mijn website', language: 'nl-NL' },
    { title: 'Hi there!', description: 'Welcome to my website', language: 'en-US' },
  ]
}
```

with the param it could become

```js
{
  translations: { title: 'Hallo wereld!', description: 'Welkom op mijn website', language: 'nl-BE' }
}
```

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._

**Reply by vanling** [reply]
@rijkvanzanten 👀 ☝️

**Reply by rijkvanzanten** [reply]
haha nou vooruit dan maar, nog even net op de valreep 😉

**Comment by northern-mindset** [comment]
Is it going to cover the case of default translation #3681? I.e. when there is no `nl-BE` item at all, and Directus will take some default (`en`?) translation item?

**Reply by rijkvanzanten** [reply]
That's the goal!

**Comment by MarvinXu** [comment]
Anyone have temporary solution for this?

**Reply by just-maik** [reply]
Query target and fallback language with the "_in" operator and then display the fallback using null-coalescence in the Frontend.
That's how we do it currently, with the extra that we define default and fallback language in directus as well. An API-side fallback would be nice though to reduce the fetched data.

**Reply by MarvinXu** [reply]
@just-maik Do you mean you have to make another request for the fallback language? Can you elaborate more? Maybe a Collection Schema can help me understand :)

**Reply by just-maik** [reply]
![image](https://github.com/directus/directus/assets/16877165/5f6b3b2d-548c-43ae-93f8-1a75f6c9ec9f)

This is the websites collection. I use it to define language related settings and match it by subdomain. Of course you could use a singleton website-settings collection.

From here it depends on your frontend stack. I personally use nuxt to load this information on the server and include it into the client bundle so I don't need to query it twice. 

Now when querying you can pass parameters. That way you make sure only your current and default languages are fetched
```gql
query GetTranslationsWithFallback($default_lang: String!, $current_lang: String!) {
	products_translations(
		filter: { languages_id: { code: { _in: [$default_lang, $current_lang] } } }
	) {
		languages_id {
			code
		}
		name
		usp_list
	}
}
```
Using this you could write a little helper function to extract the right language
```ts
import { GetTranslationsWithFallbackQuery } from "#gql";

export const getTranslation = (t: GetTranslationsWithFallbackQuery, current: string) => {
	return (
		t.products_translations.find((e) => e.languages_id?.code === current) ??
		t.products_translations[0] ??
		undefined
	);
};
```

**Reply by MarvinXu** [reply]
Thanks for your detailed reply! So this is swiching the entire translations to default language, rather than only missing field, right?

**Reply by just-maik** [reply]
Yes it is. But this depends on how you implement the helper. This is kinda hacky but I got a PoC working using this:

```ts
export function getTranslation<T extends Record<string, unknown>>(
  translations: T[],
  currentLanguage: string,
  defaultLanguage: string,
): T {
  const defaultTranslation = translations.find(
    (t) => t.language === defaultLanguage,
  );
  const currentTranslation = translations.find(
    (t) => t.language === currentLanguage,
  );

  if (!(currentTranslation && defaultTranslation)) {
    throw new Error(
      `Couldn't find translations for current language (${currentLanguage}) or default language (${defaultLanguage})`,
    );
  }

  return Object.fromEntries(
    Object.entries(currentTranslation).map(([key, value]) => [
      key,
      value || defaultTranslation[key],
    ]),
  ) as T;
}
```

**Reply by just-maik** [reply]
[PoC codesandbox ](https://codesandbox.io/p/sandbox/suspicious-gwen-3twnpj)
